### DISCUSSION TITLE
원격에서 gRPC 서버 연동 방법

### DISCUSSION THREAD

**Discussion by Card2Me** [elbakramer/koapy#38] [Q&A]
안녕하세요.

키움이 언제까지 32bit만 지원할런지.. 좀 답답하네요. 

koapy serve를 서버를 올려 놓고, 응용프로그램만 실행하는 별도의 서버에서 연결하는 방법이 있을까요?

**Comment by elbakramer** [comment]
안녕하세요.

내부적으로 gRPC 를 사용하기 때문에 관련 기능들을 활용하면
실질적으로는 말씀주신 것처럼 원격서버를 구성해서 연결하는게 가능하긴 합니다.

다만 KOAPY 가 기본적으로 로컬호스트에서 동작하도록 되어있고
말씀주신 것과 같은 구성을 그다지 적극적으로 알려주거나 드러내지 않았던건
혹시나 네크워크를 통해 발생할 수 있는 보안상의 이슈를 최대한 원천차단하기 위한 의도에서 였습니다.
특히 금전과 관련된 어플리케이션이기 때문에 문제가 발생할 경우 더욱 피해가 직접적일 것으로 생각했구요.

그렇기 때문에 이후 아래에서 말씀드릴 내용들은
충분히 보안상의 리스크를 검토하시고 그에 대한 대책을 세우셔서 사용하시길 권장드립니다.

---

지금까지는 [`KiwoomOpenApiPlusEntrypoint`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/core/KiwoomOpenApiPlusEntrypoint.py) 와 [`KiwoomOpenApiPlusTrayApplication`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusTrayApplication.py) (`koapy serve`) 인터페이스를 주로 사용하셨을건데요. 이것들에서는 위에서 말씀드린 연유로 host 를 설정할 수 있는 방법이 의도적으로 빠져있습니다.

따라서 host 를 직접 설정하기 위해서 당장은 그보다는 더 하위의 인터페이스를 사용해 직접 서버와 클라이언트를 다루셔야 하는데요. 그것들이 바로 [`KiwoomOpenApiPlusServiceClient`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusServiceClient.py) 와 [`KiwoomOpenApiPlusServiceServer`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusServiceServer.py) 입니다. 각각 [`KiwoomOpenApiPlusEntrypoint`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/core/KiwoomOpenApiPlusEntrypoint.py) 와 [`KiwoomOpenApiPlusTrayApplication`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusTrayApplication.py) 내부에서 사용되기 때문에 실질적으로 하는일은 서로 크게 다르지 않다고 보시면 됩니다.

구체적으로 각각의 객체 생성시에 네트워크와 관련해서 `host` 와 `port` 를 인자로 주면 해당 값을 직접 설정할 수 있습니다. 예를 들어 서버의 경우는 일반적으로 `host` 에 `0.0.0.0` 을 주면 모든 네트워크로부터의 연결을 허용하게 되는거고요. 클라이언트의 경우는 이후 해당 서버의 도메인이나 ip 주소를 `host` 로 주시면 그쪽으로 연결을 시도할겁니다.

서버가 외부 공용 네트워크에서의 접근이 불가능하고 내부 개인 네트워크상에서만 접근 가능하게끔 설정되었다면 여기까지 만으로도 충분할듯 싶은데요. 만약 그게 아니라면 [gRPC 의 Authentication 기능](https://grpc.io/docs/guides/auth/)을 활용하는걸 권장드립니다.

사실 현재까지는 딱히 해당 부분을 설정할 방법이 없었는데요. 이번에 https://github.com/elbakramer/koapy/commit/57c42dc8ee1170827b958807a376efdddb78e601 를 통해서 관련 credentials 를 전달하는게 가능하도록 했고 해당 수정이 0.4.2 버전에 포함되어 배포되었습니다. 앞의 커밋에 관련 기능 테스트를 위한 예시코드도 포함되어 있으니 한번 참고하시면 될 것 같습니다. 해당 예시에서는 ssl 을 쓰고 있는데, 말고 다른 다양한 방법들과 관련해서는 [gRPC Python](https://grpc.github.io/grpc/python/grpc.html) 문서의 credentials 관련 부분들도 확인해보세요.

<details>
<summary>이외에</summary>

이거는 또 별개의 이야기지만, 별도 서버에서 상시 떠있게 만드려면 아래같은 처리도 주기적으로 해줘야 될 것으로 생각되네요. 

* 매일 새벽에 진행되는 서버 점검으로 인한 키움서버와의 커넥션 종료 대응, 재연결
  * 아래같이 시도해본건 몇개 있는데 둘 다 잘 동작하진 않는거 같아요...
    * https://github.com/elbakramer/koapy/blob/83bb0d1cf7848e3a3334484c95b38ebe1c160aaf/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusTrayApplication.py#L308-L317
    * https://github.com/elbakramer/koapy/blob/83bb0d1cf7848e3a3334484c95b38ebe1c160aaf/koapy/backend/kiwoom_open_api_plus/grpc/KiwoomOpenApiPlusTrayApplication.py#L102-L148
* 자동 로그인 설정, 이후 주기적으로 자동 로그인 설정을 해제하고 OpenAPI 업데이트 확인하고 필요하다면 업데이트 처리
  * [`KiwoomOpenApiPlusVersionUpdater`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/core/KiwoomOpenApiPlusVersionUpdater.py) 를 통해서 자동 업데이트 처리 기능까진 구현했지만, 그래서 언제 어떤 주기로 확인해 줘야 할지랑 그 스케줄링을 뭘로 해줘야 할지는 잘 모르겠습니다.

저도 별도 서버에 띄워놓는 생각을 해본적은 있지만, 말씀드린 보안상의 이슈와 위와 같은 처리를 어떻게 하는게 좋을지 몰라서 일단 덮어두었는데, 혹시나 해보시고 잘 되면 알려주세요. ㅎㅎ

</details>
